src/approaches/grammar_search_invention_approach.py

```diff
@@ -396,6 +397,220 @@ def _create_grammar(grammar_name: str, dataset: Dataset,
+                                  hadd_fn: utils.HAddHeuristic,
+                                  ground_ops: Collection[_GroundSTRIPSOperator]
+                                  ) -> float:
+        score = 0
+        for i in range(len(atoms_sequence)-1):
+            atoms, next_atoms = atoms_sequence[i], atoms_sequence[i+1]
+            ground_op_demo_prob = 0. # total probability mass for demo actions
+            ground_op_total_prob = 0.  # total probability mass for all actions
+            for ground_op in ground_ops:
+                # Only care about applicable ground ops.
+                if not ground_op.preconditions.issubset(atoms):
+                    continue
+                # Compute the next state under the operator.
+                successor_atoms = atoms.copy()
+                for atom in ground_op.add_effects:
+                    successor_atoms.add(atom)
+                for atom in ground_op.delete_effects:
+                    successor_atoms.discard(atom)
+                # Compute the heuristic for the successor atoms.
+                h = hadd_fn(utils.atoms_to_tuples(successor_atoms))
+                # Compute the probability that the correct next atoms would be
+                # output under an energy-based policy.
+                k = CFG.grammar_search_lookahead_softmax_constant
+                p = np.exp(-k*h) if not np.isinf(h) else 0.
+                ground_op_total_prob += p
+                # Check whether the successor atoms match the demonstration.
+                if successor_atoms == next_atoms:
+                    ground_op_demo_prob += p
+            # If there is a demonstration state that is a dead-end under the
+            # operators, immediately return a very bad score, because planning
+            # with these operators would never be able to recover the demo.
+            if ground_op_demo_prob == 0:
+                return float("inf")
+            # Accumulate the log probability of each (state, action) in this
+            # demonstrated trajectory.
+            atoms_log_prob = np.log(ground_op_demo_prob) - \
+                             np.log(ground_op_total_prob)
+            score += -atoms_log_prob  # remember that lower is better
+        return score
+
+
+def _create_heuristic(initial_predicates: Set[Predicate],
+        atom_dataset: List[GroundAtomTrajectory],
+        candidates: Dict[Predicate, float]
+        ) -> _PredicateSearchHeuristic:
+    if CFG.grammar_search_heuristic == "prediction_error":
+        return _PredictionErrorHeuristic(initial_predicates, atom_dataset,
+                                         candidates)
+    if CFG.grammar_search_heuristic == "hadd_match":
+        return _HAddMatchHeuristic(initial_predicates, atom_dataset,
+                                   candidates)
+    if CFG.grammar_search_heuristic == "hadd_lookahead_match":
+        return _HaddLookaheadHeuristic(initial_predicates, atom_dataset,
```

Comment:
Naming inconsistency: `_HAddMatchHeuristic` vs `_HaddLookaheadHeuristic`, should use `HAdd` everywhere